src/term-spec.js

```diff
@@ -566,10 +712,138 @@ declare export class VariableDeclaration extends Term {
+  kind: any; // 'any' | 'number' | 'object' | 'boolean' | 'string' | 'symbol' |
+             // 'this' | 'void' | 'undefined' | 'null' | 'never'
+}
+
+// type FunctionOrConstructorTypeNode = FunctionTypeNode | ConstructorTypeNode
+declare export class SignatureTypeNode extends TypeNode {
+  typeParameters?: TypeParameterDeclaration[];
+  params: FormalParameters;
+  type?: TypeNode;
+}
+// example: `(i: number, s: string) => boolean`
+declare export class FunctionTypeNode extends SignatureTypeNode {}
+// example: `new (i: number, s: string) => SomeClass`
+declare export class ConstructorNodeType extends SignatureTypeNode {}
+
+// type TypeReferenceType = TypeReferenceNode | ExpressionWithTypeArguments;
+
+declare export class TypeReferenceNode extends TypeNode {
+  typeName: any; // Identifier
+  typeArguments?: TypeNode[];
+}
+
+// example: `SomeMapWithGenericClassValues["key"]<number, string>`
+declare export class ExpressionWithTypeArguments extends TypeNode {
+  parent?: HeritageClause;
+  /*
+   * In TypeScript, the "destination" class/interface of a heritage clause is
+   * allowed to be an arbitrary `LeftHandSideExpression`` which would perhaps
+   * best translate into `AssignmentTarget`. However, in the original version
+   * of this spec, the superclass is an arbitrary Expression, so here we retain
+   * that permissiveness.
+   */
+  expression: Expression;
+  typeArguments?: TypeNode[];
+}
+
+// example 1:
+//   `something is string` in a boolean-returning function declared as
+//   `function isString(something: any): something is string`
+// example 2:
+//   `this is string` in a boolean-returning class method declared as
+//   `isString(): this is string`
+declare export class TypePredicateNode extends TypeNode {
+  parameterName: any; //  Identifier | 'this'
+  type: TypeNode;
+}
+
+// example: `typeof someVariable`
+declare export class TypeQueryNode extends TypeNode {
+  exprName: any; // Identifier
+}
+
+// example: `{ readonly someKey: number }`
+declare export class TypeLiteralNode extends TypeNode {
+  members: (TypeElement | IndexSignatureDeclaration)[];
+}
+
+// example: `string[]`
+declare export class ArrayTypeNode extends TypeNode {
+  elementType: TypeNode;
+}
+
+// example: `[string, number]`
+declare export class TupleTypeNode extends TypeNode {
+  elementTypes: TypeNode[];
+}
+
+// example: `number | string`
+declare export class UnionTypeNode extends TypeNode {
+  types: TypeNode[];
+}
+
+// example: `number & string`
+declare export class IntersectionTypeNode extends TypeNode {
+  types: TypeNode[];
+}
+
+// example: `(number)`
+declare export class ParenthesizedTypeNode extends TypeNode {
+  type: TypeNode;
+}
+
+// example: `keyof { a: number, b: string }`
+declare export class TypeOperatorNode extends TypeNode {
+  operator: any; // 'keyof'
+  type: TypeNode;
+}
+
+// example:
+//   `SomeClass[number]`
+//   where
+//   `class SomeClass { [index: number]: string }`
```

Comment:
The comment above does not mention that `SomeClass[number]` would then essentially resolve to the `string` type.